# Data Understanding After Cleaning

After performing the cleaning, we check again some basic information to show that all the data in correct in the sintax and the semantic

In [1]:
# Autoreload allows the notebook to dynamically load code: if we update some helper functions *outside* of the notebook, we do not need to reload the notebook.
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd
import numpy as np
import re
import unicodedata

In [3]:
csv_file = "../data/merged_dataset.csv"
dataset = pd.read_csv(csv_file)
dataset.head()

_url_cyclist         name_cyclist  birth_year  weight  height  \
0       gerard-vianen       Gerard  Vianen      1944.0     NaN     NaN   
1        jos-huysmans        Jos  Huysmans      1941.0     NaN     NaN   
2     romain-furniere     Romain  Furniere      1943.0     NaN     NaN   
3         willy-donie         Willy  Donie      1945.0     NaN     NaN   
4  frans-melckenbeeck  Frans  Melckenbeeck      1940.0     NaN     NaN   

   nationality                     _url_race         name_race  points  \
0  Netherlands  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
1      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
2      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
3      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
4      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   

   uci_points  ...  startlist_quality        date  position  \
0         NaN  ...                372  1970-04-25        23   
1         NaN  ...                372  1970-04-25        18   
2         NaN  ...                372  1970-04-25        31   
3         NaN  ...                372  1970-04-25        27   
4         NaN  ...                372  1970-04-25         8   

              cyclist cyclist_age  mostly_tarmac             cyclist_team  \
0       gerard-vianen        26.0          False   uae-team-emirates-2018   
1        jos-huysmans        29.0          False          team-monex-2005   
2     romain-furniere        27.0          False                      NaN   
3         willy-donie        25.0          False                      NaN   
4  frans-melckenbeeck        30.0          False  team-qhubeka-assos-2021   

   delta  year   stage  
0   52.0  1970  result  
1   52.0  1970  result  
2   52.0  1970  result  
3   52.0  1970  result  
4   33.0  1970  result  

[5 rows x 23 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 589739 entries, 0 to 589738
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   _url_cyclist       589739 non-null  object 
 1   name_cyclist       589739 non-null  object 
 2   birth_year         589739 non-null  float64
 3   weight             480007 non-null  float64
 4   height             480007 non-null  float64
 5   nationality        589739 non-null  object 
 6   _url_race          589739 non-null  object 
 7   name_race          589739 non-null  object 
 8   points             589739 non-null  float64
 9   uci_points         422184 non-null  float64
 10  length             589739 non-null  float64
 11  climb_total        573157 non-null  float64
 12  profile            573157 non-null  float64
 13  startlist_quality  589739 non-null  int64  
 14  date               589739 non-null  object 
 15  position           589739 non-null  int64  
 16  cy

In [5]:
dataset.describe()

birth_year         weight         height         points  \
count  589739.000000  480007.000000  480007.000000  589739.000000   
mean     1977.683221      68.362072     179.699850      89.225897   
std        11.537102       6.182201       6.269585      54.415806   
min      1933.000000      48.000000     154.000000      18.000000   
25%      1970.000000      64.000000     175.000000      50.000000   
50%      1979.000000      68.000000     180.000000      80.000000   
75%      1986.000000      73.000000     184.000000     100.000000   
max      2004.000000      94.000000     204.000000     350.000000   

          uci_points         length    climb_total        profile  \
count  422184.000000  589739.000000  573157.000000  573157.000000   
mean       72.048275  166776.290800    2319.964102       2.497300   
std       121.041689   64543.334848    1353.290485       1.478509   
min         6.000000    1000.000000       2.000000       1.000000   
25%        16.000000  152500.000000    1331.000000       1.000000   
50%        20.000000  178200.000000    2269.000000       2.000000   
75%       100.000000  203500.000000    3274.000000       4.000000   
max       800.000000  338000.000000    6974.000000       5.000000   

       startlist_quality       position    cyclist_age          delta  \
count      589739.000000  589739.000000  589739.000000  589739.000000   
mean         1101.182822      74.217408      28.485834     430.755887   
std           380.588233      48.405261       3.855921    1011.486288   
min           115.000000       0.000000      13.000000       0.000000   
25%           844.000000      32.000000      26.000000      11.000000   
50%           988.000000      70.000000      28.000000     158.000000   
75%          1309.000000     112.000000      31.000000     626.000000   
max          2047.000000     209.000000      56.000000   46380.000000   

                year  
count  589739.000000  
mean     2006.169054  
std        11.476909  
min      1970.000000  
25%      1999.000000  
50%      2008.000000  
75%      2015.000000  
max      2023.000000

## Checks of cyclists data

Now we start from the data related to the initial cyclist dataset

### Check on '_url' data

We start considering the `_url` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in _url column: ' + str(dataset['_url'].isnull().sum())
      + ' (' + str(round(dataset['_url'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in _url column:')
url_counts = dataset['_url'].value_counts()
print(url_counts)

Total number of null values in _url column: 0 (0.0%)

Count occurrences of each value in _url column:
_url
ward-vanhoof            1
bruno-surra             1
gerard-rue              1
jan-maas                1
nathan-van-hooydonck    1
                       ..
stian-remme             1
scott-davies            1
joost-van-leijen        1
chad-haga               1
willy-moonen            1
Name: count, Length: 6134, dtype: int64


We have lots different values, but no null values.

## Check on 'name' data

Now we consider the `name` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in name column: ' + str(dataset['name'].isnull().sum())
        + ' (' + str(round(dataset['name'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in name column:')
name_counts = dataset['name'].value_counts()
print(name_counts)   

Total number of null values in name column: 0 (0.0%)

Count occurrences of each value in name column:
name
Sergio  Domínguez       2
Alberto  Fernández      2
Jesús  López            2
Antonio  Cabello        2
Alessandro  Pozzi       2
                       ..
Juan José  Martínez     1
Iñigo  Elosegui         1
Paolo  Alberati         1
Jackson  Rodríguez      1
Jean-Philippe  Dojwa    1
Name: count, Length: 6127, dtype: int64


We have lots of different values, but no null values.

Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# For each data, check if 'name' object contains any number
# Check if the 'name' column contains any number
invalid_rows = dataset[dataset['name'].str.contains(r'\d')]

print('Number of rows with numbers in the name: ' + str(len(invalid_rows)))

Number of rows with numbers in the name: 0


## Check on 'birth_year' data

Now we consider the `birth_year` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in birth_year column: ' + str(dataset['birth_year'].isnull().sum())
        + ' (' + str(round(dataset['birth_year'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in birth_year column:')
birth_year_counts = dataset['birth_year'].value_counts()
print(birth_year_counts)

Total number of null values in birth_year column: 13 (0.21%)

Count occurrences of each value in birth_year column:
birth_year
1964.0    145
1962.0    141
1970.0    140
1974.0    138
1980.0    133
         ... 
1937.0      4
1934.0      2
1938.0      2
1933.0      1
1936.0      1
Name: count, Length: 71, dtype: int64


We have different values, and a few null values

Since we have different values, we check if every value is sintatically correct

In [ ]:
# Get rows where 'birth_year' does not end with '.0'
invalid_rows = dataset[~dataset['birth_year'].astype(str).str.endswith('.0')].dropna(subset=['birth_year'])

print('Number of invalid birth_year: ' + str(len(invalid_rows)))

Number of invalid birth_year: 0


In this block we check if the `birth year` value is not in the form 'nnnn' and if it is not in the form '19nn' or '20nn'

In [ ]:
# Remove '.0' from 'birth_year' and check if it's a valid year in the form '19nn' or '20nn'
def is_valid_year(birth_year):
    # Convert to string and remove '.0'
    year = str(birth_year).replace('.0', '')
    # Check if the year is a digit and starts with '19' or '20'
    return year.isdigit() and (year.startswith('19') or year.startswith('20'))

# Filter out rows where 'birth_year' is not null
invalid_rows = dataset[dataset['birth_year'].notnull()]

# Apply the validation function across the 'birth_year' column
invalid_rows = invalid_rows[~invalid_rows['birth_year'].apply(is_valid_year)]

print('Number of invalid birth_year: ' + str(len(invalid_rows)))


Number of invalid birth_year: 0


Check the races where the `birth_year` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'birth_year' column
dataset['birth_year'].describe()

count    6121.000000
mean     1974.071884
std        15.535834
min      1933.000000
25%      1962.000000
50%      1974.000000
75%      1987.000000
max      2004.000000
Name: birth_year, dtype: float64

The birth years are between 1933 and 2004, so we can say that no extreme values are present (in the cyclism context).

## Check on 'weight' data

Now we consider the `weight` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in weight column: ' + str(dataset['weight'].isnull().sum())
        + ' (' + str(round(dataset['weight'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in weight column:')
weight_counts = dataset['weight'].value_counts()
print(weight_counts)

Total number of null values in weight column: 3056 (49.82%)

Count occurrences of each value in weight column:
weight
70.0    272
68.0    219
65.0    193
67.0    177
72.0    169
69.0    162
73.0    146
63.0    140
66.0    139
64.0    137
74.0    135
62.0    131
75.0    128
71.0    125
60.0     98
61.0     90
78.0     86
77.0     67
58.0     64
76.0     63
80.0     53
59.0     49
79.0     30
82.0     26
55.0     25
81.0     22
83.0     20
57.0     20
56.0     19
85.0     10
53.0      7
52.0      6
84.0      6
54.0      4
51.0      4
90.0      4
87.0      3
88.0      3
63.5      2
89.0      2
50.0      2
58.5      2
86.0      2
71.5      1
48.0      1
91.0      1
67.5      1
66.5      1
78.1      1
77.5      1
74.5      1
81.4      1
62.5      1
93.0      1
73.5      1
79.5      1
65.1      1
92.0      1
94.0      1
Name: count, dtype: int64


We have different values, but a lot of null values. Also, we see that every value is sintatically correct

Check the races where the `weight` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'weight' column
dataset['weight'].describe()

count    3078.000000
mean       68.658739
std         6.348183
min        48.000000
25%        64.000000
50%        69.000000
75%        73.000000
max        94.000000
Name: weight, dtype: float64

The weights are between 48 and 94 kg, so we can say that no extreme values are present (in the cyclism context).

## Check on 'height' data

Now we consider the `height` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in height column: ' + str(dataset['height'].isnull().sum())
        + ' (' + str(round(dataset['height'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in height column:')
height_counts = dataset['height'].value_counts()
print(height_counts)

Total number of null values in height column: 2991 (48.76%)

Count occurrences of each value in height column:
height
180.0    277
178.0    226
183.0    193
181.0    181
175.0    169
182.0    165
185.0    161
176.0    154
184.0    152
179.0    137
177.0    133
174.0    129
173.0    120
186.0    107
190.0     97
170.0     90
187.0     85
172.0     80
188.0     77
171.0     67
189.0     48
169.0     46
191.0     37
192.0     34
168.0     24
167.0     23
193.0     22
164.0     20
194.0     17
195.0     13
165.0     13
196.0      7
197.0      6
166.0      6
198.0      4
160.0      4
162.0      3
159.0      3
161.0      2
199.0      2
163.0      2
154.0      1
204.0      1
155.0      1
158.0      1
202.0      1
157.0      1
200.0      1
Name: count, dtype: int64


We have different values, but a lot of null values. Also, we see that every value is sintatically correct

Check the races where the `height` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'height' column
dataset['height'].describe()

count    3143.000000
mean      179.815145
std         6.443447
min       154.000000
25%       175.000000
50%       180.000000
75%       184.000000
max       204.000000
Name: height, dtype: float64

The heights are between 154 and 204 cm, so we can say that no extreme values are present (in the cyclism context).

## Check on 'nationality' data

Now we consider the `nationality` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in nationality column: ' + str(dataset['nationality'].isnull().sum())
        + ' (' + str(round(dataset['nationality'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in nationality column:')
nationality_counts = dataset['nationality'].value_counts()
print(nationality_counts)

Total number of null values in nationality column: 1 (0.02%)

Count occurrences of each value in nationality column:
nationality
Italy                 1029
Spain                  872
Belgium                869
France                 741
Netherlands            380
                      ... 
Dominican Republic       1
Liechtenstein            1
Zimbabwe                 1
Puerto Rico              1
Hongkong                 1
Name: count, Length: 72, dtype: int64


We have different values, and just one null value.

Since we have a lot of different values, we check if every value is sintatically correct

In this block we check if there are `nationality` values that contains any character that is a letter

In [ ]:
# Check if there are 'nationality' values that contains any number, without the null values
invalid_rows = dataset[dataset['nationality'].notnull() & dataset['nationality'].str.contains(r'\d')]

print('Number of rows with numbers in nationality: ' + str(len(invalid_rows)))

Number of rows with numbers in nationality: 0


# Checks of races data

In [ ]:
# Autoreload allows the notebook to dynamically load code: if we update some helper functions *outside* of the notebook, we do not need to reload the notebook.
%load_ext autoreload
%autoreload 2

In [ ]:
import pandas as pd
import numpy as np
import re
import unicodedata

We load the dataset from a CSV file and display the first few rows to get an initial understanding of the data. This helps us verify that the data has been loaded correctly and gives us a glimpse of its structure and contents.

In [ ]:
csv_file = "../data/races.csv"
dataset = pd.read_csv(csv_file)
dataset.head()

_url            name  points  uci_points    length  \
0  tour-de-france/1978/stage-6  Tour de France   100.0         NaN  162000.0   
1  tour-de-france/1978/stage-6  Tour de France   100.0         NaN  162000.0   
2  tour-de-france/1978/stage-6  Tour de France   100.0         NaN  162000.0   
3  tour-de-france/1978/stage-6  Tour de France   100.0         NaN  162000.0   
4  tour-de-france/1978/stage-6  Tour de France   100.0         NaN  162000.0   

   climb_total  profile  startlist_quality  average_temperature  \
0       1101.0      1.0               1241                  NaN   
1       1101.0      1.0               1241                  NaN   
2       1101.0      1.0               1241                  NaN   
3       1101.0      1.0               1241                  NaN   
4       1101.0      1.0               1241                  NaN   

                  date  position           cyclist  cyclist_age  is_tarmac  \
0  1978-07-05 04:02:24         0        sean-kelly         22.0       True   
1  1978-07-05 04:02:24         1  gerrie-knetemann         27.0       True   
2  1978-07-05 04:02:24         2    rene-bittinger         24.0       True   
3  1978-07-05 04:02:24         3    joseph-bruyere         30.0       True   
4  1978-07-05 04:02:24         4  sven-ake-nilsson         27.0       True   

   is_cobbled  is_gravel                          cyclist_team  delta  
0       False      False  vini-ricordi-pinarello-sidermec-1986    0.0  
1       False      False                           norway-1987    0.0  
2       False      False                                   NaN    0.0  
3       False      False              navigare-blue-storm-1993    0.0  
4       False      False                            spain-1991    0.0

Create a dataset without the personal information of the cyclists, taking only one row per race

In [ ]:
# Delete 'position', 'cyclist', 'cyclist_age', 'cyclist_team' and 'delta' columns
races_info = dataset.drop(columns=['position', 'cyclist', 'cyclist_age', 'cyclist_team', 'delta'])

# For each row in 'races_info', take only the year-month-day part of 'date' (delete the time)
races_info['date'] = races_info['date'].str.split(' ').str[0]

# Eliminate duplicates
races_info = races_info.drop_duplicates()

# Display the first rows of the dataset
races_info.head()

_url                        name  points  \
0        tour-de-france/1978/stage-6              Tour de France   100.0   
106    vuelta-a-espana/2016/stage-14             Vuelta a España    80.0   
271     tour-de-france/2019/stage-21              Tour de France   100.0   
426  volta-a-catalunya/1999/prologue  Volta Ciclista a Catalunya    50.0   
545      tour-de-france/2022/stage-9              Tour de France   100.0   

     uci_points    length  climb_total  profile  startlist_quality  \
0           NaN  162000.0       1101.0      1.0               1241   
106       100.0  196000.0       5575.0      5.0                821   
271       120.0  128000.0        781.0      1.0               1699   
426         NaN    8100.0          NaN      NaN                804   
545       120.0  192900.0       3743.0      3.0               1551   

     average_temperature        date  is_tarmac  is_cobbled  is_gravel  
0                    NaN  1978-07-05       True       False      False  
106                  NaN  2016-09-03       True       False      False  
271                  NaN  2019-07-28       True       False      False  
426                  NaN  1999-06-17       True       False      False  
545                 24.0  2022-07-10       True       False      False

Create dataset from the union of the cyclists and the races data 

In [ ]:
# Create union of two datasets, merging them considering the url of the cyclist
dataset_cyclists = pd.read_csv("../data/cyclists.csv")
dataset_races = pd.read_csv("../data/races.csv")
merged_dataset = pd.merge(dataset_cyclists, dataset_races, left_on='_url', right_on='cyclist', how='inner')

# Modify name column of the cyclist url in '_url_cyclist', and name column of the race url in '_url_race'
merged_dataset = merged_dataset.rename(columns={'_url_x': '_url_cyclist', '_url_y': '_url_race'})
# Modify name column of the cyclist name in 'name_cyclist', and name column of the race name in 'name_race'
merged_dataset = merged_dataset.rename(columns={'name_x': 'name_cyclist', 'name_y': 'name_race'})
# Take only the year-month-day part of 'date' (delete the time)
merged_dataset['date'] = merged_dataset['date'].str.split(' ').str[0]

merged_dataset.head()


_url_cyclist  name_cyclist  birth_year  weight  height nationality  \
0  bruno-surra  Bruno  Surra      1964.0     NaN     NaN       Italy   
1   gerard-rue   Gérard  Rué      1965.0    74.0   182.0      France   
2   gerard-rue   Gérard  Rué      1965.0    74.0   182.0      France   
3   gerard-rue   Gérard  Rué      1965.0    74.0   182.0      France   
4   gerard-rue   Gérard  Rué      1965.0    74.0   182.0      France   

                      _url_race        name_race  points  uci_points  ...  \
0  vuelta-a-espana/1989/stage-1  Vuelta a España    80.0         NaN  ...   
1   tour-de-france/1997/stage-2   Tour de France   100.0         NaN  ...   
2   tour-de-france/1990/stage-1   Tour de France   100.0         NaN  ...   
3   tour-de-france/1992/stage-7   Tour de France   100.0         NaN  ...   
4   tour-de-france/1990/stage-9   Tour de France   100.0         NaN  ...   

   average_temperature        date  position      cyclist  cyclist_age  \
0                  NaN  1989-04-24       110  bruno-surra         25.0   
1                  NaN  1997-07-07       132   gerard-rue         32.0   
2                  NaN  1990-07-01        66   gerard-rue         25.0   
3                  NaN  1992-07-11        35   gerard-rue         27.0   
4                  NaN  1990-07-09        41   gerard-rue         25.0   

  is_tarmac  is_cobbled is_gravel  cyclist_team  delta  
0      True       False     False           NaN   15.0  
1      True       False     False  denmark-1991    0.0  
2      True       False     False   france-1978  635.0  
3      True       False     False   france-1978   65.0  
4      True       False     False   france-1978   37.0  

[5 rows x 24 columns]

## Initial Info

Now we provide a concise summary of the DataFrame, including the number of non-null entries, data types of each column, and memory usage. It helps us quickly identify missing values and understand the overall structure of the dataset.

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 589865 entries, 0 to 589864
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   _url                 589865 non-null  object 
 1   name                 589865 non-null  object 
 2   points               589388 non-null  float64
 3   uci_points           251086 non-null  float64
 4   length               589865 non-null  float64
 5   climb_total          442820 non-null  float64
 6   profile              441671 non-null  float64
 7   startlist_quality    589865 non-null  int64  
 8   average_temperature  29933 non-null   float64
 9   date                 589865 non-null  object 
 10  position             589865 non-null  int64  
 11  cyclist              589865 non-null  object 
 12  cyclist_age          589752 non-null  float64
 13  is_tarmac            589865 non-null  bool   
 14  is_cobbled           589865 non-null  bool   
 15  is_gravel        

Also, we generates a descriptive statistics for numerical columns in the DataFrame. It includes metrics such as count, mean, standard deviation, minimum, and maximum values, as well as the 25th, 50th, and 75th percentiles. This summary helps us understand the distribution and central tendency of the data.

In [ ]:
dataset.describe()

points     uci_points         length    climb_total  \
count  589388.000000  251086.000000  589865.000000  442820.000000   
mean       89.221635      74.601547  166776.180584    2330.469215   
std        54.435330     100.947962   64545.605664    1375.710722   
min        18.000000       6.000000    1000.000000       2.000000   
25%        50.000000      16.000000  152500.000000    1309.000000   
50%        80.000000      60.000000  178200.000000    2255.000000   
75%       100.000000     100.000000  203500.000000    3273.000000   
max       350.000000     800.000000  338000.000000    6974.000000   

             profile  startlist_quality  average_temperature       position  \
count  441671.000000      589865.000000         29933.000000  589865.000000   
mean        2.611611        1101.161178            21.731768      74.219491   
std         1.491741         380.586928             5.884761      48.404023   
min         1.000000         115.000000            10.000000       0.000000   
25%         1.000000         844.000000            17.000000      32.000000   
50%         2.000000         988.000000            22.000000      70.000000   
75%         4.000000        1309.000000            26.000000     112.000000   
max         5.000000        2047.000000            36.000000     209.000000   

         cyclist_age          delta  
count  589752.000000  589865.000000  
mean       28.486208     418.292794  
std         3.855631     842.961596  
min        13.000000   -6906.000000  
25%        26.000000      10.000000  
50%        28.000000     156.000000  
75%        31.000000     624.000000  
max        56.000000   61547.000000

## Check on '_url' data

We start considering the `_url` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in _url column: ' + str(dataset['_url'].isnull().sum())
      + ' (' + str(round(dataset['_url'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in _url column:')
url_counts = dataset['_url'].value_counts()
print(url_counts)

Total number of null values in _url column: 0 (0.0%)

Count occurrences of each value in _url column:
_url
tour-de-france/1986/stage-1     210
tour-de-france/1986/prologue    210
tour-de-france/1987/prologue    207
giro-d-italia/2011/stage-2      206
tour-de-france/1987/stage-1     206
                               ... 
il-lombardia/1972/result          1
giro-d-italia/2001/stage-14       1
tour-de-suisse/1972/stage-2       1
tour-de-suisse/1972/stage-7       1
tour-de-suisse/1972/stage-6b      1
Name: count, Length: 5281, dtype: int64


We have lots of different values, but no null values

In this block we check if there are `_url` values that are not in the form name/year/stage

In [ ]:
# Split url by / in name, year and stage
url_split = dataset['_url'].str.split('/', expand=True)
# Check null elements in url_split[0], url_split[1] and url_split[2], and if url_split[1] contains only digits
invalid_rows = dataset[url_split[0].isnull() | url_split[1].isnull() | url_split[2].isnull() | ~url_split[1].str.isdigit()]
print('Number of invalid URLs: ' + str(len(invalid_rows)))

Number of invalid URLs: 0


In [ ]:
#TODO: Chiedere sulle gare dove ci sono un solo o pochi partecipanti

url_counts = dataset['_url'].value_counts()

# From url_counts, get the urls where the number of occurrence is less than 2
print(url_counts[url_counts < 6])

_url
tour-de-romandie/1992/stage-6     5
dauphine/1983/prologue            5
tour-de-romandie/1989/stage-1     5
tirreno-adriatico/1996/stage-7    5
volta-a-catalunya/1992/stage-2    5
                                 ..
il-lombardia/1972/result          1
giro-d-italia/2001/stage-14       1
tour-de-suisse/1972/stage-2       1
tour-de-suisse/1972/stage-7       1
tour-de-suisse/1972/stage-6b      1
Name: count, Length: 146, dtype: int64


## Check on 'name' data

Now we consider the `name` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in name column: ' + str(dataset['name'].isnull().sum())
      + ' (' + str(round(dataset['name'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\n\nCount occurrences of each value in name column:')
name_counts = dataset['name'].value_counts()
print(name_counts)

Total number of null values in name column: 0 (0.0%)


Count occurrences of each value in name column:
name
Tour de France                     145500
Giro d'Italia                       95581
Vuelta a España                     89222
Tour de Suisse                      33682
Paris - Nice                        32362
                                    ...  
E3 Saxo Classic                       101
E3 BinckBank Classic                   99
E3 Prijs Vlaanderen - Harelbeke        98
Clásica Ciclista San Sebastian         84
Clásica San Sebastián                  52
Name: count, Length: 61, dtype: int64


We have different values, but no null values

Since it's small, we print all the values

In [ ]:
# Print all the names that appear, alphabetically ordered
print(name_counts.index.sort_values())

Index(['Amstel Gold Race', 'Clasica Ciclista San Sebastian',
       'Clásica Ciclista San Sebastian', 'Clásica Ciclista San Sebastián',
       'Clásica San Sebastián', 'Criterium du Dauphiné',
       'Criterium du Dauphiné Libére', 'Critérium du Dauphiné',
       'Critérium du Dauphiné Libéré', 'Donostia San Sebastian Klasikoa',
       'Dwars door België / À travers la Belgique', 'Dwars door Vlaanderen',
       'Dwars door Vlaanderen - A travers la Flandre ME',
       'Dwars door Vlaanderen / A travers la Flandre',
       'Dwars door Vlaanderen / A travers la Flandre ME',
       'E3 BinckBank Classic', 'E3 Harelbeke', 'E3 Prijs Vlaanderen',
       'E3 Prijs Vlaanderen - Harelbeke', 'E3 Saxo Bank Classic',
       'E3 Saxo Classic', 'E3-Prijs Harelbeke', 'Giro d'Italia',
       'Giro di Lombardia', 'Gran Camiño', 'Grand Prix Cycliste de Montréal',
       'Grand Prix Cycliste de Quebec', 'Grand Prix Cycliste de Québec',
       'Il Lombardia', 'Itzulia Basque Country', 'La Flèche Wallonne'

For each url, check if all the `name` values are the same


In [ ]:
# Group by '_url' and calculate the number of unique values in the 'name' column
name_uniques = dataset.groupby('_url')['name'].nunique(dropna=False)
# Filter the URLs with more than one unique name
multiple_names_urls = name_uniques[name_uniques > 1].index

print('Number of URLs with more than one unique name: ' + str(len(multiple_names_urls)))

Number of URLs with more than one unique name: 0


In this block we are checking if there are `name` values that contains any incorrect numbers

In [ ]:
# Get rows wehere 'name' value contains any number, except for names containing 'E3' (there are some races with E3 in the name)
invalid_rows = dataset[dataset['name'].str.contains(r'\d') & ~dataset['name'].str.contains('E3')]
print('Number of invalid names: ' + str(len(invalid_rows)))

Number of invalid names: 0


## Check on 'points' data

Now we consider the `point` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in points column: ' + str(dataset['points'].isnull().sum())
      + ' (' + str(round(dataset['points'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in points column:')
point_counts = dataset['points'].value_counts()
print(point_counts)

Total number of null values in points column: 477 (0.08%)

Count occurrences of each value in points column:
points
80.0     198878
50.0     186102
100.0    141706
275.0     22299
225.0     19536
125.0      5992
30.0       4313
350.0      3917
70.0       3299
75.0       1963
20.0        792
18.0        292
35.0        183
150.0       116
Name: count, dtype: int64


We have not a lot of values, and few null values. Also, we see that every value is sintatically correct

For each url, check if all the `name` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'points' column
points_uniques = dataset.groupby('_url')['points'].nunique(dropna=False)
# Filter the URLs with more than one unique points
multiple_points_urls = points_uniques[points_uniques > 1].index

print('Number of URLs with more than one unique points: ' + str(len(multiple_points_urls)))

Number of URLs with more than one unique points: 0


We check the urls where `points` is null

In [ ]:
# Get unique data based on '_url' and 'points'
unique_data = dataset.drop_duplicates(subset=['_url', 'points'])
# Get rows where 'points' is null
rows = unique_data[unique_data['points'].isnull()]

print('Number of urls with null points: ' + str(len(rows)))
print('\nURLs of the rows with null points:')
print(rows['_url'])

Number of urls with null points: 4

URLs of the rows with null points:
156755    vuelta-a-espana/1994/stage-5
461300    tour-de-france/1986/stage-19
517517    tour-de-france/1988/prologue
561313    tour-de-france/2019/stage-19
Name: _url, dtype: object


## Check on 'uci_points' data

Now we consider the `uci_points` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in uci_points column: ' + str(dataset['uci_points'].isnull().sum())
      + ' (' + str(round(dataset['uci_points'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in uci_points column:')
uci_point_counts = dataset['uci_points'].value_counts()
print(uci_point_counts)

Total number of null values in uci_points column: 338779 (57.43%)

Count occurrences of each value in uci_points column:
uci_points
100.0    47640
6.0      43390
16.0     41103
60.0     39317
20.0     21303
120.0    20785
50.0     13266
500.0     6096
40.0      4102
210.0     3449
180.0     3103
400.0     2571
300.0     2057
14.0       792
600.0      675
800.0      514
200.0      338
80.0       328
10.0       148
402.0      109
Name: count, dtype: int64


We have different values, but a lot of null values. Also, we see that every value is sintatically correct

For each url, check if all the `uci_points` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'uci_points' column
uci_points_uniques = dataset.groupby('_url')['uci_points'].nunique(dropna=False)
# Filter the URLs with more than one unique uci_points
multiple_uci_points_urls = uci_points_uniques[uci_points_uniques > 1].index

print('Number of URLs with more than one unique uci_points: ' + str(len(multiple_uci_points_urls)))

Number of URLs with more than one unique uci_points: 0


We check the urls where `uci_points` is null

In [ ]:
# Get unique data based on '_url' and 'uci_points'
unique_data = dataset.drop_duplicates(subset=['_url', 'uci_points'])
# Get rows where 'uci_points' is null
rows = unique_data[unique_data['uci_points'].isnull()]

print('Number of rows with null uci_points: ' + str(len(rows)))
print('\nURLs of the rows with null uci_points:')
print(rows['_url'])

Number of rows with null uci_points: 3682

URLs of the rows with null uci_points:
0             tour-de-france/1978/stage-6
426       volta-a-catalunya/1999/prologue
866          tour-de-france/1978/stage-14
1075      volta-a-catalunya/1981/stage-2b
1084             paris-nice/1994/stage-8b
                       ...               
588250              gp-quebec/2010/result
588712       tour-de-france/1982/stage-21
588837     tirreno-adriatico/1993/stage-3
589397          paris-roubaix/2000/result
589463            paris-nice/1976/stage-2
Name: _url, Length: 3682, dtype: object


We check rows where `point` is null but not `uci_point`

In [ ]:
# Get unique data based on '_url' and 'uci_points'
unique_data = dataset.drop_duplicates(subset=['_url', 'uci_points'])
# Get rows where 'uci_points' is not null, but 'points' is null
rows = unique_data[unique_data['uci_points'].notnull() & unique_data['points'].isnull()]

print('Number of rows with null uci_points: ' + str(len(rows)))

Number of rows with null uci_points: 0


## Check on 'length' data

Now we consider the `length` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in length column: ' + str(dataset['length'].isnull().sum())
      + ' (' + str(round(dataset['length'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in length column:')
length_counts = dataset['length'].value_counts()
print(length_counts)

Total number of null values in length column: 0 (0.0%)

Count occurrences of each value in length column:
length
177000.0    5039
170000.0    4717
195000.0    4413
200000.0    4401
178000.0    4286
            ... 
4500.0         5
107000.0       5
123100.0       5
142700.0       5
2000.0         3
Name: count, Length: 1280, dtype: int64


We have a lot of values, but no null values

For each url, check if all the `length` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'length' column
length_uniques = dataset.groupby('_url')['length'].nunique(dropna=False)
# Filter the URLs with more than one unique length
multiple_length_urls = length_uniques[length_uniques > 1].index

print('Number of URLs with more than one unique length: ' + str(len(multiple_length_urls)))

Number of URLs with more than one unique length: 0


Since we have a lot of values, we check if every value is sintatically correct.

In [ ]:
# Check that every value is a digit
dataset['length'].dtype

dtype('float64')

In [ ]:
# Get rows where 'length' does not end with '.0'
invalid_rows = dataset[~dataset['length'].astype(str).str.endswith('.0')]
                                
print('Number of invalid lengths: ' + str(len(invalid_rows)))
for index, row in invalid_rows.iterrows():
    print(row['_url'], row['length'])

Number of invalid lengths: 954
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/20

Check the races where the `length` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'length' column
dataset['length'].describe()

count    589865.000000
mean     166776.180584
std       64545.605664
min        1000.000000
25%      152500.000000
50%      178200.000000
75%      203500.000000
max      338000.000000
Name: length, dtype: float64

The lenghts are between 1000 and 338000 m, so we can say that no extreme values are present.

## Check on 'climb_total' data

Now we consider the `climb_total` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in climb_total column: ' + str(dataset['climb_total'].isnull().sum())
      + ' (' + str(round(dataset['climb_total'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in climb_total column:')
climb_total_counts = dataset['climb_total'].value_counts()
print(climb_total_counts)

Total number of null values in climb_total column: 147045 (24.93%)

Count occurrences of each value in climb_total column:
climb_total
3500.0    3762
2500.0    3261
4000.0    3029
3000.0    2938
5000.0    2592
          ... 
3742.0       9
2525.0       9
1176.0       8
1903.0       5
3128.0       1
Name: count, Length: 2117, dtype: int64


We have a lot of different values, and a lot of null values

For each url, check if all the `climb_total` values are the same


In [ ]:
# Group by '_url' and calculate the number of unique values in the 'climb_total' column
climb_total_uniques = dataset.groupby('_url')['climb_total'].nunique(dropna=False)
# Filter the URLs with more than one unique climb_total
multiple_climb_total_urls = climb_total_uniques[climb_total_uniques > 1].index

print('Number of URLs with more than one unique climb_total: ' + str(len(multiple_climb_total_urls)))

Number of URLs with more than one unique climb_total: 0


Since we have a lot of values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['climb_total'].dtype

dtype('float64')

In [ ]:
# Get rows where 'climb_total' does not end with '.0'
invalid_rows = dataset[~dataset['climb_total'].astype(str).str.endswith('.0')].dropna(subset=['climb_total'])

print('Number of invalid climb_total: ' + str(len(invalid_rows)))

Number of invalid climb_total: 0


Check the races where the `climb_total` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'climb_total' column
dataset['climb_total'].describe()

count    442820.000000
mean       2330.469215
std        1375.710722
min           2.000000
25%        1309.000000
50%        2255.000000
75%        3273.000000
max        6974.000000
Name: climb_total, dtype: float64

The climb values are between 2 and 6974 m, so we can say that no extreme values are present.

## Check on 'profile' data

Now we consider the `profile` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in profile column: ' + str(dataset['profile'].isnull().sum())
      + ' (' + str(round(dataset['profile'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in profile column:')
profile_counts = dataset['profile'].value_counts()
print(profile_counts)

Total number of null values in profile column: 148194 (25.12%)

Count occurrences of each value in profile column:
profile
1.0    131344
2.0    128269
5.0     88203
3.0     50844
4.0     43011
Name: count, dtype: int64


We have few different values, but a lot of null values. Also, we see that every value is sintatically correct

For each url, check if all the `profile` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'profile' column
profile_uniques = dataset.groupby('_url')['profile'].nunique(dropna=False)
# Filter the URLs with more than one unique profile
multiple_profile_urls = profile_uniques[profile_uniques > 1].index

print('Number of URLs with more than one unique profile: ' + str(len(multiple_profile_urls)))

Number of URLs with more than one unique profile: 0


## Check on 'startlist_quality' data

Now we consider the `startlist_quality` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in startlist_quality column: ' + str(dataset['startlist_quality'].isnull().sum())
      + ' (' + str(round(dataset['startlist_quality'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in startlist_quality column:')
startlist_quality_counts = dataset['startlist_quality'].value_counts()
print(startlist_quality_counts)

Total number of null values in startlist_quality column: 0 (0.0%)

Count occurrences of each value in startlist_quality column:
startlist_quality
971     8279
1812    7807
1872    7317
1612    7255
920     6715
        ... 
455        5
438        5
394        5
228        3
544        3
Name: count, Length: 697, dtype: int64


We have different values, but no null values.

For each url, check if all the `startlist_quality` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'startlist_quality' column
startlist_quality_uniques = dataset.groupby('_url')['startlist_quality'].nunique(dropna=False)
# Filter the URLs with more than one unique startlist_quality
multiple_startlist_quality_urls = startlist_quality_uniques[startlist_quality_uniques > 1].index

print('Number of URLs with more than one unique startlist_quality: ' + str(len(multiple_startlist_quality_urls)))

Number of URLs with more than one unique startlist_quality: 0


Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['startlist_quality'].dtype

dtype('int64')

## Check on 'average_temperature' data

Now we consider the `average_temperature` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in average_temperature column: ' + str(dataset['average_temperature'].isnull().sum())
      + ' (' + str(round(dataset['average_temperature'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in average_temperature column:')
average_temperature_counts = dataset['average_temperature'].value_counts()
print(average_temperature_counts)

Total number of null values in average_temperature column: 559932 (94.93%)

Count occurrences of each value in average_temperature column:
average_temperature
23.0    2343
22.0    2099
20.0    2055
24.0    1829
25.0    1789
14.0    1499
21.0    1463
26.0    1440
18.0    1435
13.0    1386
15.0    1295
16.0    1213
17.0    1170
28.0    1149
19.0    1080
29.0    1038
31.0    1026
27.0     990
30.0     769
11.0     731
32.0     730
12.0     548
36.0     265
35.0     167
33.0     165
34.0     133
10.0     126
Name: count, dtype: int64


We have different values, and almost all the values are null. Also, we cas see that every value is sintatically correct

For each url, check if all the `average_temperature` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'average_temperature' column
average_temperature_uniques = dataset.groupby('_url')['average_temperature'].nunique(dropna=False)
# Filter the URLs with more than one unique average_temperature
multiple_average_temperature_urls = average_temperature_uniques[average_temperature_uniques > 1].index

print('Number of URLs with more than one unique average_temperature: ' + str(len(multiple_average_temperature_urls)))

Number of URLs with more than one unique average_temperature: 0


## Check on 'date' data

Now we consider the `date` column, and check the number of null values and the count the occurrences of each unique value. We do this considerig the merged dataset where we don't have time value of the date

In [ ]:
print('Total number of null values in date column: ' + str(merged_dataset['date'].isnull().sum())
      + ' (' + str(round(merged_dataset['date'].isnull().sum() / len(merged_dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in date column:')
date_counts = merged_dataset['date'].value_counts()
print(date_counts)

Total number of null values in date column: 0 (0.0%)

Count occurrences of each value in date column:
date
1998-06-18    407
1987-07-05    404
2006-05-15    379
2006-05-16    377
2023-02-25    376
             ... 
1972-06-23      1
1972-06-21      1
1972-06-17      1
1972-06-22      1
2001-06-02      1
Name: count, Length: 4708, dtype: int64


We have different values, but no null values.

For each url, check if all the `date` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'date' column
date_uniques = dataset.groupby('_url')['date'].nunique(dropna=False)
# Filter the URLs with more than one unique date
multiple_date_urls = date_uniques[date_uniques > 1].index

print('Number of URLs with more than one unique date (format yyyy-mm-dd hh:mm:ss, in the races dataset): ' + str(len(multiple_date_urls)))

# Group by '_url' and calculate the number of unique values in the 'date' column
date_uniques = merged_dataset.groupby('_url_race')['date'].nunique(dropna=False)
# Filter the URLs with more than one unique date
multiple_date_urls = date_uniques[date_uniques > 1].index

print('Number of URLs with more than one unique date (format yyyy-mm-dd, in the merged dataset): ' + str(len(multiple_date_urls)))

Number of URLs with more than one unique date (format yyyy-mm-dd hh:mm:ss, in the races dataset): 4777
Number of URLs with more than one unique date (format yyyy-mm-dd, in the merged dataset): 0


Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# Check if there are any 'date' values in the format yyyy-mm-dd hh:mm:ss (in the races dataset)
invalid_rows = dataset[~dataset['date'].str.match(r'\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}')]
print('Number of invalid dates (format yyyy-mm-dd hh:mm:ss, in the races dataset): ' + str(len(invalid_rows)))

# Check if there are any 'date' values not in the format yyyy-mm-dd (in the merged dataset)
invalid_rows = merged_dataset[~merged_dataset['date'].str.match(r'\d{4}-\d{2}-\d{2}')]
print('Number of invalid dates (format yyyy-mm-dd, in the merged dataset): ' + str(len(invalid_rows)))

Number of invalid dates (format yyyy-mm-dd hh:mm:ss, in the races dataset): 0
Number of invalid dates (format yyyy-mm-dd, in the merged dataset): 0


Check if the year is the same in both the `_url` and the `date`

In [ ]:
# Split _url by / into name, year, and stage
url_split = dataset['_url'].str.split('/', expand=True) # expand=True to return a DataFrame
# Extract the year from the date column (assuming format yyyy-mm-dd hh:mm:ss)
date_year = dataset['date'].str[:4]
# Compare the year in the _url (from the second part of the split) with the year in the date
mismatched_years = dataset[(url_split[1] != date_year)]

# Print the number of rows where the year does not match
print(f"Number of rows where the year in the url does not match the year in the date: {len(mismatched_years)}")


Number of rows where the year in the url does not match the year in the date: 0


## Check on 'position' data

Now we consider the `position` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in position column: ' + str(dataset['position'].isnull().sum())
      + ' (' + str(round(dataset['position'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in position column:')
position_counts = dataset['position'].value_counts()
print(position_counts)

Total number of null values in position column: 0 (0.0%)

Count occurrences of each value in position column:
position
0      5281
1      5275
2      5273
3      5267
4      5255
       ... 
205       5
206       3
207       2
208       2
209       2
Name: count, Length: 210, dtype: int64


We have different values, and no null values.

Since we have different values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['position'].dtype

dtype('int64')

For each url, check if there are all the `position` values

In [ ]:
# Function to check if the positions are from 0 to the max one after the other
def check_positions(positions):
    return np.array_equal(np.sort(positions), np.arange(positions.max() + 1))

# Apply the function to the dataset
invalid_urls = dataset.groupby('_url')['position'].apply(lambda x: not check_positions(x))

# Stampa gli '_url' che non rispettano la condizione
print('Number of URLs with invalid positions: ' + str(len(invalid_urls[invalid_urls])))


Number of URLs with invalid positions: 0


## Check on 'cyclist' data

Now we consider the `cyclist` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in cyclist column: ' + str(dataset['cyclist'].isnull().sum())
      + ' (' + str(round(dataset['cyclist'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in cyclist column:')
cyclist_counts = dataset['cyclist'].value_counts()
print(cyclist_counts)

Total number of null values in cyclist column: 0 (0.0%)

Count occurrences of each value in cyclist column:
cyclist
matteo-tosatto         959
alejandro-valverde     942
luis-leon-sanchez      899
imanol-erviti          883
haimar-zubeldia        883
                      ... 
timothy-vangheel         1
matthias-friedemann      1
stefan-rucker            1
john-brouwer             1
stijn-ennekens           1
Name: count, Length: 6095, dtype: int64


We have lots of different values, but no null values.

We check if all the cyclists are different in the same race

In [ ]:
# For each url, check if the a cyclist appears more than once
url_cyclist_count = dataset.groupby('_url')['cyclist'].value_counts()
invalid_entries = url_cyclist_count[url_cyclist_count > 1]

print('Number of URLs with a cyclist appearing more than once: ' + str(len(invalid_entries.index.get_level_values(0))))

# Estrai gli _url e i ciclisti che compaiono più volte
for (url, cyclist), count in invalid_entries.items():
    print(f"URL: {url}, Cyclist: {cyclist}, Count: {count}")

Number of URLs with a cyclist appearing more than once: 123
URL: dauphine/2005/stage-1, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-2, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-3, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-4, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-5, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-6, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2005/stage-7, Cyclist: ivan-gutierrez, Count: 2
URL: dauphine/2012/stage-3, Cyclist: david-moncoutie, Count: 2
URL: itzulia-basque-country/2000/stage-1, Cyclist: alberto-david-fernandez, Count: 2
URL: itzulia-basque-country/2000/stage-2, Cyclist: alberto-david-fernandez, Count: 2
URL: itzulia-basque-country/2000/stage-2, Cyclist: david-vazquez, Count: 2
URL: itzulia-basque-country/2000/stage-2, Cyclist: miguel-angel-pena, Count: 2
URL: itzulia-basque-country/2000/stage-3, Cyclist: miguel-angel-pena, Count: 2
URL: itzulia-basque-country/2000/stage-4

Considering the two datasets, we check if all the cyclists in cyclists.csv are in races.csv, and vice versa.

In [ ]:
# Check if there are any 'cyclist' values in the races dataset that are not in the cyclists dataset
invalid_rows = dataset[~dataset['cyclist'].isin(dataset_cyclists['_url'])]

print('Number of cyclists with no info: ' + str(len(invalid_rows)))

Number of cyclists with no info: 0


In [ ]:
# Check if there are any '_url' values in the cyclists dataset that are not in the races dataset
invalid_rows = dataset_cyclists[~dataset_cyclists['_url'].isin(dataset['cyclist'])]

print('Number of cyclists that are not in any race: ' + str(len(invalid_rows)))
for index, row in invalid_rows.iterrows():
    print(row['_url'], row['name'])

Number of cyclists that are not in any race: 39
jean-michel-thilloy Jean-Michel  Thilloy
gert-van-brabant Gert Van Brabant
roman-bronis Roman  Broniš
oleg-grishkine Oleg  Grishkine
eddy-torrekens Eddy  Torrekens
philipp-ludescher Philipp  Ludescher
nicolas-liboreau Nicolas  Liboreau
gino-primo Gino  Primo
luca-braidot Luca  Braidot
tanner-putt Tanner  Putt
matteo-di-serafino Matteo Di Serafino
jeanot-deriemaecker Jeanot  Deriemaecker
dorian-de-maeght Dorian De Maeght
martin-gilbert Martin  Gilbert
eric-schoefs Eric  Schoefs
silvere-ackermann Silvère  Ackermann
franck-morelle Franck  Morelle
christian-mager Christian  Mager
rikkie-matthijssens Rikkie  Matthijssens
marat-ganeev Marat  Ganeev
bas-tietema Bas  Tietema
mattia-viel Mattia  Viel
hiroki-nishimura Hiroki  Nishimura
christophe-premont Christophe  Premont
kurt-van-landeghem Kurt van Landeghem
lenaic-olivier Lénaïc  Olivier
arturo-gravalos-lopez Arturo  Grávalos
morten-hegreberg Morten  Hegreberg
rik-claeys Rik  Claeys
pascal-duez

## Check on 'cyclist_age' data

Now we consider the `cyclist_age` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in cyclist_age column: ' + str(dataset['cyclist_age'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_age'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in cyclist_age column:')
cyclist_age_counts = dataset['cyclist_age'].value_counts()
print(cyclist_age_counts)

Total number of null values in cyclist_age column: 113 (0.02%)

Count occurrences of each value in cyclist_age column:
cyclist_age
27.0    58897
26.0    57921
28.0    56213
25.0    54346
29.0    52616
30.0    46860
24.0    43252
31.0    40827
32.0    35063
23.0    30009
33.0    27972
34.0    21965
35.0    16413
22.0    13531
36.0    12111
37.0     8141
38.0     4703
21.0     3733
39.0     2470
40.0     1127
20.0      842
41.0      307
42.0      218
43.0       96
19.0       90
45.0       16
44.0       11
56.0        1
13.0        1
Name: count, dtype: int64


We have different values, and just a few of null values. Also, we see that every value is sintatically correct

Check the races where the `cyclist_age` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'cyclist_age' column
dataset['cyclist_age'].describe()

count    589752.000000
mean         28.486208
std           3.855631
min          13.000000
25%          26.000000
50%          28.000000
75%          31.000000
max          56.000000
Name: cyclist_age, dtype: float64

The ages are between 13 and 56 cm, so we can say that no extreme values are present (in the cyclism context).

## Check on 'is_tarmac' data

Now we consider the `is_tarmac` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in is_tarmac column: ' + str(dataset['is_tarmac'].isnull().sum())
      + ' (' + str(round(dataset['is_tarmac'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in is_tarmac column:')
is_tarmac_counts = dataset['is_tarmac'].value_counts()
print(is_tarmac_counts)

Total number of null values in is_tarmac column: 0 (0.0%)

Count occurrences of each value in is_tarmac column:
is_tarmac
True     536042
False     53823
Name: count, dtype: int64


We have two different values, and no null values. Also, we see that every value is sintatically correct

For each url, check if all the `is_tarmac` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'is_tarmac' column
is_tarmac_uniques = dataset.groupby('_url')['is_tarmac'].nunique(dropna=False)
# Filter the URLs with more than one unique is_tarmac
multiple_is_tarmac_urls = is_tarmac_uniques[is_tarmac_uniques > 1].index

print('Number of URLs with more than one unique is_tarmac: ' + str(len(multiple_is_tarmac_urls)))

Number of URLs with more than one unique is_tarmac: 0


## Check on 'is_cobbled' data

Now we consider the `is_cobbled` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in is_cobbled column: ' + str(dataset['is_cobbled'].isnull().sum())
      + ' (' + str(round(dataset['is_cobbled'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in is_cobbled column:')
is_cobbled_counts = dataset['is_cobbled'].value_counts()
print(is_cobbled_counts)

Total number of null values in is_cobbled column: 0 (0.0%)

Count occurrences of each value in is_cobbled column:
is_cobbled
False    589865
Name: count, dtype: int64


We have one value, and no null values. Also, we see that the value is sintatically correct

## Check on 'is_gravel' data

Now we consider the `is_gravel` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in is_gravel column: ' + str(dataset['is_gravel'].isnull().sum())
      + ' (' + str(round(dataset['is_gravel'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in is_gravel column:')
is_gravel_counts = dataset['is_gravel'].value_counts()
print(is_gravel_counts)

Total number of null values in is_gravel column: 0 (0.0%)

Count occurrences of each value in is_gravel column:
is_gravel
False    589865
Name: count, dtype: int64


We have one value, and no null values. Also, we see that the value is sintatically correct

## Check on 'cyclist_team' data

Now we consider the `cyclist_team` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
# Print total number of null values in 'delta' column, and the percentage of null values (float with two decimal digits after the comma)
print('Total number of null values in cyclist_team column: ' + str(dataset['cyclist_team'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_team'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in cyclist_team column:')
cyclist_team_counts = dataset['cyclist_team'].value_counts()
print(cyclist_team_counts)

Total number of null values in cyclist_team column: 159161 (26.98%)

Count occurrences of each value in cyclist_team column:
cyclist_team
liberty-seguros-wurth-team-2005     8869
roompot-nederlandse-loterij-2018    8773
chazal-vetta-mbk-1993               8094
kondor-1979                         7895
kazakhstan-2019                     7701
                                    ... 
atala-ofmega-1988                   1259
finland-2016                        1236
south-africa-1993                   1174
denmark-2003                         216
quickstep-innergetic-2009              3
Name: count, Length: 91, dtype: int64


We have different values, and a lot of null values.

Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# Check if every 'cyclist_team' value matches the required format
invalid_teams = dataset[dataset['cyclist_team'].notnull() & ~dataset['cyclist_team'].astype(str).str.match(r'.+-\d{4}')]

print('Number of invalid rows: ' + str(len(invalid_teams)))

Number of invalid rows: 0


## Check on 'delta' data

Now we consider the `delta` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in delta column: ' + str(dataset['delta'].isnull().sum())
        + ' (' + str(round(dataset['delta'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in delta column:')
delta_counts = dataset['delta'].value_counts()
print(delta_counts)

Total number of null values in delta column: 0 (0.0%)

Count occurrences of each value in delta column:
delta
0.0       120546
2.0         3700
5.0         3353
3.0         3178
4.0         3036
           ...  
2453.0         1
2357.0         1
2889.0         1
2718.0         1
2527.0         1
Name: count, Length: 2836, dtype: int64


We have lots of different values, but no null values.

Since we have a lot of different values, we check if every value is sintatically correct.

In [ ]:
# Check that every value is a digit
dataset['delta'].dtype

dtype('float64')

In [ ]:
# Check if there are any 'delta' values that do not end with '.0'
invalid_rows = dataset[~dataset['delta'].astype(str).str.endswith('.0')]

print('Number of invalid deltas: ' + str(len(invalid_rows)))

Number of invalid deltas: 0


In [ ]:
# Check if there are any 'delta' values that are negative
invalid_rows = dataset[dataset['delta'] < 0]

print('Number of negative deltas: ' + str(len(invalid_rows)))
for index, row in invalid_rows.iterrows():
    print(row['_url'], row['delta'])

Number of negative deltas: 86
vuelta-a-espana/1992/stage-19 -2635.0
vuelta-a-espana/1992/stage-19 -2638.0
vuelta-a-espana/1992/stage-19 -2541.0
vuelta-a-espana/1992/stage-19 -2542.0
vuelta-a-espana/1992/stage-19 -2545.0
vuelta-a-espana/1992/stage-19 -2546.0
vuelta-a-espana/1992/stage-19 -2550.0
vuelta-a-espana/1992/stage-19 -2560.0
vuelta-a-espana/1992/stage-19 -2564.0
vuelta-a-espana/1992/stage-19 -2567.0
vuelta-a-espana/1992/stage-19 -2567.0
vuelta-a-espana/1992/stage-19 -2574.0
vuelta-a-espana/1992/stage-19 -2469.0
vuelta-a-espana/1992/stage-19 -2473.0
vuelta-a-espana/1992/stage-19 -2475.0
vuelta-a-espana/1992/stage-19 -2477.0
vuelta-a-espana/1992/stage-19 -2479.0
vuelta-a-espana/1992/stage-19 -2481.0
vuelta-a-espana/1992/stage-19 -2482.0
vuelta-a-espana/1992/stage-19 -2485.0
vuelta-a-espana/1992/stage-19 -2486.0
vuelta-a-espana/1992/stage-19 -2486.0
vuelta-a-espana/1992/stage-19 -2487.0
vuelta-a-espana/1992/stage-19 -2500.0
vuelta-a-espana/1992/stage-19 -2504.0
vuelta-a-espana/1992

Check if following the `positon` order, the delta is ordered too

In [ ]:
# Grouped by '_url', check if the 'delta' value is greater or equal than the previous one (the one in the previous row, except for the first one)
invalid_urls = dataset.groupby('_url')['delta'].apply(lambda x: (x.shift() > x).any())

# Print the number of URLs with invalid 'delta' values
print('Number of URLs with invalid deltas: ' + str(len(invalid_urls[invalid_urls])))
for url in invalid_urls[invalid_urls].index:
    print(url)

Number of URLs with invalid deltas: 346
dauphine/2006/stage-2
dauphine/2007/stage-7
dauphine/2008/stage-2
dauphine/2008/stage-3
dauphine/2008/stage-6
dauphine/2008/stage-7
dauphine/2009/stage-4
dauphine/2009/stage-6
dauphine/2011/stage-2
dauphine/2015/stage-2
dauphine/2016/stage-4
dauphine/2017/stage-1
dauphine/2022/stage-1
dauphine/2022/stage-2
dauphine/2023/stage-1
dauphine/2023/stage-3
dauphine/2023/stage-5
dwars-door-vlaanderen/2003/result
e3-harelbeke/1976/result
e3-harelbeke/2023/result
giro-d-italia/1991/stage-10
giro-d-italia/1992/stage-3
giro-d-italia/1994/stage-8
giro-d-italia/1995/stage-14
giro-d-italia/1995/stage-17
giro-d-italia/1995/stage-5
giro-d-italia/1995/stage-8
giro-d-italia/1997/stage-13
giro-d-italia/1997/stage-3
giro-d-italia/1999/stage-7
giro-d-italia/2002/stage-2
giro-d-italia/2002/stage-3
giro-d-italia/2002/stage-4
giro-d-italia/2003/stage-1
giro-d-italia/2005/prologue
giro-d-italia/2005/stage-1
giro-d-italia/2005/stage-15
giro-d-italia/2005/stage-2
giro-d-ita